src/DataStructures/VectorImpl.hpp

```diff
@@ -0,0 +1,452 @@
+// IWYU pragma: no_include <blaze/math/PaddingFlag.h>
+// IWYU pragma: no_include <blaze/math/traits/AddTrait.h>
+// IWYU pragma: no_include <blaze/math/traits/DivTrait.h>
+// IWYU pragma: no_include <blaze/math/traits/MultTrait.h>
+// IWYU pragma: no_include <blaze/math/traits/SubTrait.h>
+// IWYU pragma: no_include <blaze/system/TransposeFlag.h>
+// IWYU pragma: no_include <blaze/math/traits/BinaryMapTrait.h>
+// IWYU pragma: no_include <blaze/math/traits/UnaryMapTrait.h>
+// IWYU pragma: no_include <blaze/math/typetraits/TransposeFlag.h>
+
+// IWYU pragma: no_forward_declare blaze::DenseVector
+// IWYU pragma: no_forward_declare blaze::UnaryMapTrait
+// IWYU pragma: no_forward_declare blaze::BinaryMapTrait
+// IWYU pragma: no_forward_declare blaze::IsVector
+// IWYU pragma: no_forward_declare blaze::TransposeFlag
+
+namespace detail {
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief Base class template for various DataVector and related types
+ *
+ * \details The VectorImpl class is the generic parent class for Vectors
+ * representing collections of related function values, such as DataVectors
+ * for contiguous data over a computational domain, or a representation of
+ * fourier data.
+ *
+ * The VectorImpl does not itself support any particular mathematical
+ * operations on the contained values. The VectorImpl template class and the
+ * macros defined in VectorImpl.hpp assist in the construction of various
+ * derived classes supporting a chosen set of methematical operations.
+ *
+ */
+template <typename T, typename VECTOR_TYPE>
+class VectorImpl
+    : public PointerVector<T, blaze::unaligned, blaze::unpadded,
+                           blaze::defaultTransposeFlag, VECTOR_TYPE> {
+  /// \cond HIDDEN_SYMBOLS
+  static constexpr void private_asserts() noexcept {
+    static_assert(
+        std::is_nothrow_move_constructible<VectorImpl<T, VECTOR_TYPE>>::value,
+        "Missing move semantics");
+  }
+  /// \endcond
+ public:
+  using value_type = T;
+  using allocator_type = std::allocator<value_type>;
+  using size_type = size_t;
+  using difference_type = std::ptrdiff_t;
+  using BaseType = PointerVector<T, blaze::unaligned, blaze::unpadded,
+                                 blaze::defaultTransposeFlag, VECTOR_TYPE>;
+  static constexpr bool transpose_flag = blaze::defaultTransposeFlag;
+
+  using ElementType = T;
+  using TransposeType = VectorImpl<T, VECTOR_TYPE>;
+  using CompositeType = const VectorImpl<T, VECTOR_TYPE>&;
+
+  using BaseType::operator[];
+  using BaseType::begin;
+  using BaseType::cbegin;
+  using BaseType::cend;
+  using BaseType::data;
+  using BaseType::end;
+  using BaseType::size;
+
+  // @{
+  // Upcast to `BaseType`
+  const BaseType& operator~() const noexcept {
+    return static_cast<const BaseType&>(*this);
+  }
+  BaseType& operator~() noexcept { return static_cast<BaseType&>(*this); }
+    // @}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wshadow" /*ignore shadowing of size*/
```

Comment:
I'm confused:
1.  how there's shadowing that wasn't there before
2. and why we would choose to pop warnings instead of just change the name